server/timestamp/timestamp.go

```diff
@@ -0,0 +1,114 @@
+package timestamp
+
+import (
+	"encoding/json"
+	"github.com/endophage/gotuf/data"
+	"github.com/endophage/gotuf/signed"
+	"time"
+
+	"github.com/docker/notary/server/storage"
+)
+
+func GetTimestampKey(gun string, store storage.MetaStore, crypto signed.CryptoService) (*data.TUFKey, error) {
+	cipher, public, err := store.GetTimestampKey(gun)
+	if err != nil {
+		if _, ok := err.(*storage.ErrNoKey); ok {
+			key, err := crypto.Create("timestamp")
+			if err != nil {
+				return nil, err
+			}
+			err = store.SetTimestampKey(gun, key.Cipher(), []byte(key.Public()))
+			if err != nil {
+				if _, ok := err.(*storage.ErrTimestampKeyExists); ok {
+					cipher, public, err := store.GetTimestampKey(gun)
+					if err != nil {
+						return nil, err
+					}
+					return data.NewTUFKey(cipher, string(public), ""), nil
+				}
+				return nil, err
+			}
+			return &key.TUFKey, nil
+		}
+	}
+	return data.NewTUFKey(cipher, string(public), ""), nil
+}
+
+func GetTimestamp(gun string, store storage.MetaStore, signer *signed.Signer) ([]byte, error) {
+	d, err := store.GetCurrent(gun, "timestamp")
+	if err != nil {
+		if _, ok := err.(*storage.ErrNotFound); !ok {
+			// If we received an ErrNotFound, we're going to
+			// generate the first timestamp, any other error
+			// should be returned here
+			return nil, err
+		}
+	}
+	var ts *data.SignedTimestamp
+	if d != nil {
+		err := json.Unmarshal(d, ts)
+		if err != nil {
+			return nil, err
+		}
+		if !timestampExpired(ts) {
+			return d, nil
+		}
+	}
+	sgnd, version, err := createTimestamp(gun, ts, store, signer)
+	if err != nil {
+		return nil, err
+	}
+	out, err := json.Marshal(sgnd)
```

Comment:
I'm slowly working on cleaning up a lot of this stuff. Latest push fixes all the casting from []byte to string and back for the key values. Still working out how to make some of the interfaces more intelligent to abstract away the json pieces.